File: adapters/django/src/fieldseal_django/codec.py
"""Value <-> bytes, for the inner field's Python value (docs/12 §2, spec §3.6).

The core encrypts bytes. A Django field holds a `str`, an `int`, a `Decimal`,
a `date`. Something has to bridge those, and *what* it is is a security
decision rather than a convenience one.

**Never pickle.** Rails shipped `Marshal` as the serializer for encrypted
attributes, which turns any attacker who can write a ciphertext -- or any
operator restoring a doctored backup -- into remote code execution on
decrypt (`docs/04` §8). The codec here is fixed, non-executing, and refuses
types it does not know rather than reaching for a general serializer.

**The rendering is spec §3.6's, not Django's.** Until G25 (#123) this module
rendered through `get_prep_value()` and `str()`, which is CPython's text form
rather than any specified one: `Decimal("1.5")` and `Decimal("1.50")` became
different bytes and so different blind-index values, `True` became `b"True"`,
and a `float` rendered `1e+16` where the other adapter wrote
`10000000000000000`. Every inner field now maps to one of §3.6's eight logical
types (`logical_type`), the value goes through the inner field's own
`get_prep_value` / `to_python` for Django's input coercion, and the bytes are
§3.6's canonical rendering of the result. Reading is exactly as strict: bytes
that are not the canonical rendering are refused, never coerced.
"""

from __future__ import annotations

import datetime as dt
import math
import re
from decimal import Decimal
from typing import TYPE_CHECKING, Any

from django.db import models

from .errors import FieldsealNotSupported

if TYPE_CHECKING:  # pragma: no cover - typing only
    from django.db.models import Field

#: The inner field classes each §3.6 logical type is reached through, checked
#: in this order. Order matters twice: `DateTimeField` subclasses `DateField`,
#: and `BooleanField` must win before anything numeric could claim it.
_MAPPING: tuple[tuple[type[Any], str], ...] = (
    (models.BinaryField, "bytes"),
    (models.BooleanField, "boolean"),
    (models.DateTimeField, "datetime"),
    (models.DateField, "date"),
    (models.DecimalField, "decimal"),
    (models.FloatField, "float"),
    (models.IntegerField, "int"),
    (models.CharField, "string"),
    (models.TextField, "string"),
)


def logical_type(field: Field[Any, Any]) -> str | None:
    """The spec §3.6 logical type `field` holds, or None if §3.6 pins none.

    `UUIDField`, `TimeField`, `DurationField`, `JSONField`,
    `GenericIPAddressField` and every other type return None, and
    `Encrypted()` refuses them at declaration: a refused column is a schema
    change, a column rendered by `str()` is a backfill once another language
    reads it differently.
    """
    for cls, name in _MAPPING:
        if isinstance(field, cls):
            return name
    return None


def to_bytes(field: Field[Any, Any], value: Any) -> bytes:
    """Serialize `value` for encryption under spec §3.6."""
    if value is None:
        raise FieldsealNotSupported(
            "None cannot be encrypted: a NULL column is indistinguishable "
            "from an absent value to every reader, so the adapter stores "
            "NULL as NULL rather than encrypting a placeholder that would "
            "claim the row has a value it does not"
        )
    kind = _require_type(field)
    try:
        return _RENDER[kind](field, value)
    except FieldsealNotSupported:
        raise
    except Exception as e:  # noqa: BLE001 - Django's coercion failed: typed refusal
        raise FieldsealNotSupported(
            f"{type(field).__name__} (§3.6 `{kind}`) could not take "
            f"{type(value).__name__}: {e}"
        ) from e


def from_bytes(field: Field[Any, Any], raw: bytes) -> Any:
    """Inverse of `to_bytes`: refuse anything but the canonical rendering.

    The decrypted value is authentic -- key, context and commitment all
    verified -- so a refusal here means the bytes were written by something
    that does not render §3.6, or the column's declared type changed after the
    row was written. Either way a coerced value would hide it.
    """
    kind = _require_type(field)
    value = _PARSE[kind](raw)
    if kind in ("bytes", "string"):
        return value
    # A custom inner field may post-process its own type; the built-in ones
    # return the value unchanged.
    return field.to_python(value)


# -- write --------------------------------------------------------------------

def _render_string(field: Field[Any, Any], value: Any) -> bytes:
    prepared = field.get_prep_value(value)
    if not isinstance(prepared, str):
        raise FieldsealNotSupported(
            f"{type(field).__name__}.get_prep_value returned "
            f"{type(prepared).__name__}, not str")
    try:
        return prepared.encode("utf-8")
    except UnicodeEncodeError as e:
        raise FieldsealNotSupported(
            "the string contains an unpaired surrogate, which is not a Unicode "
            "scalar value and has no UTF-8 encoding (spec §3.6)") from e


def _render_bytes(field: Field[Any, Any], value: Any) -> bytes:
    if isinstance(value, bytes | bytearray | memoryview):
        return bytes(value)
    raise FieldsealNotSupported(
        f"a BinaryField takes bytes, not {type(value).__name__}")


def _render_int(field: Field[Any, Any], value: Any) -> bytes:
    prepared = field.get_prep_value(value)
    if type(prepared) is not int:
        raise FieldsealNotSupported(f"not an int: {type(prepared).__name__}")
    return str(prepared).encode("ascii")


def _render_decimal(field: Field[Any, Any], value: Any) -> bytes:
    if isinstance(value, float):
        # DecimalField.to_python would round a float to the field's context;
        # §3.6 refuses rather than rounds.
        raise FieldsealNotSupported(
            "a float is not a decimal: pass Decimal or str so no binary64 "
            "rounding happens before the value is encrypted")
    prepared = field.get_prep_value(value)
    if not isinstance(prepared, Decimal):
        raise FieldsealNotSupported(f"not a Decimal: {type(prepared).__name__}")
    return render_decimal(prepared)


def _render_float(field: Field[Any, Any], value: Any) -> bytes:
    prepared = field.get_prep_value(value)
    if not isinstance(prepared, float):
        raise FieldsealNotSupported(f"not a float: {type(prepared).__name__}")
    return render_float(prepared)


def _render_boolean(field: Field[Any, Any], value: Any) -> bytes:
    prepared = field.get_prep_value(value)
    if not isinstance(prepared, bool):
        raise FieldsealNotSupported(f"not a bool: {type(prepared).__name__}")
    return b"true" if prepared else b"false"


def _render_date(field: Field[Any, Any], value: Any) -> bytes:
    if isinstance(value, dt.datetime):
        # DateField.to_python would convert an aware datetime to the default
        # time zone's date and drop the time; §3.6 refuses rather than truncates.
        raise FieldsealNotSupported(
            "a datetime is not a date: pass datetime.date, so the calendar "
            "date stored is the one the application chose")
    prepared = field.to_python(value)
    if not isinstance(prepared, dt.date) or isinstance(prepared, dt.datetime):
        raise FieldsealNotSupported(f"not a date: {type(prepared).__name__}")
    return prepared.isoformat().encode("ascii")


def _render_datetime(field: Field[Any, Any], value: Any) -> bytes:
    # `to_python`, not `get_prep_value`: under USE_TZ the latter makes a naive
    # value aware in the default zone, which is exactly the assumption §3.6
    # forbids.
    prepared = field.to_python(value)
    if not isinstance(prepared, dt.datetime):
        raise FieldsealNotSupported(f"not a datetime: {type(prepared).__name__}")
    if prepared.tzinfo is None or prepared.utcoffset() is None:
        raise FieldsealNotSupported(
            "a naive datetime has no instant: spec §3.6 refuses rather than "
            "assuming a time zone. Pass an aware datetime")
    try:
        u = prepared.astimezone(dt.UTC)
    except OverflowError as e:
        raise FieldsealNotSupported("outside years 0001-9999 in UTC") from e
    return (f"{u.year:04d}-{u.month:02d}-{u.day:02d}T{u.hour:02d}:"
            f"{u.minute:02d}:{u.second:02d}.{u.microsecond:06d}Z").encode("ascii")


def render_decimal(d: Decimal) -> bytes:
    """Spec §3.6 canonical decimal. From `as_tuple()`, never `normalize()`,
    which rounds to the 28-digit context precision."""
    if not d.is_finite():
        raise FieldsealNotSupported(f"a decimal must be finite, not {d}")
    sign, digit_tuple, exp = d.as_tuple()
    assert isinstance(exp, int)
    digits = "".join(map(str, digit_tuple)).lstrip("0")
    if not digits:
        return b"0"
    stripped = digits.rstrip("0")
    exp += len(digits) - len(stripped)
    if exp >= 0:
        body = stripped + "0" * exp
    else:
        point = len(stripped) + exp
        body = (stripped[:point] + "." + stripped[point:] if point > 0
                else "0." + "0" * -point + stripped)
    return (("-" if sign else "") + body).encode("ascii")


def render_float(x: float) -> bytes:
    """Spec §3.6: ECMA-262 Number::toString(x), with negative zero as `-0`.

    `repr` is CPython's shortest round-trip digit string; this lays those
    digits out by ECMAScript's rules, which is where CPython differs (it
    writes `1e+16` and `1e-07` where ECMAScript writes `10000000000000000`
    and `1e-7`).
    """
    if not math.isfinite(x):
        raise FieldsealNotSupported(f"a float must be finite, not {x!r}")
    if x == 0:
        return b"-0" if math.copysign(1.0, x) < 0 else b"0"
    if x < 0:
        return b"-" + render_float(-x)
    mant, _, e = repr(x).partition("e")
    ip, _, fp = mant.partition(".")
    raw = (ip + fp).lstrip("0")
    s = raw.rstrip("0")
    k = len(s)
    n = k + int(e or 0) - len(fp) + (len(raw) - k)
    if k <= n <= 21:
        out = s + "0" * (n - k)
    elif 0 < n <= 21:
        out = s[:n] + "." + s[n:]
    elif -6 < n <= 0:
        out = "0." + "0" * -n + s
    else:
        out = ((s if k == 1 else s[0] + "." + s[1:])
               + "e" + ("+" if n > 0 else "-") + str(abs(n - 1)))
    return out.encode("ascii")


# -- read ---------------------------------------------------------------------

_INT = re.compile(r"0|-?[1-9][0-9]*")
_DECIMAL = re.compile(r"-?(0|[1-9][0-9]*)(\.[0-9]*[1-9])?")
_FLOAT = re.compile(r"-?[0-9]+(\.[0-9]+)?(e[+-][0-9]+)?")
_DATE = re.compile(r"[0-9]{4}-[0-9]{2}-[0-9]{2}")
_DATETIME = re.compile(
    r"[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}\.[0-9]{6}Z")


def _refuse(kind: str, raw: bytes) -> FieldsealNotSupported:
    shown = raw[:40]
    return FieldsealNotSupported(
        f"the decrypted value {shown!r} is not spec §3.6's canonical `{kind}` "
        "rendering. The envelope was authentic, so this is not tampering: the "
        "row was written by something that does not render §3.6, or the "
        "column's declared type changed after it was written. Changing the "
        "type is a new plaintext encoding and needs a backfill, not an edit.")


def _ascii(kind: str, raw: bytes) -> str:
    try:
        return raw.decode("ascii")
    except UnicodeDecodeError:
        raise _refuse(kind, raw) from None


def _parse_string(raw: bytes) -> str:
    try:
        return raw.decode("utf-8")
    except UnicodeDecodeError as e:
        raise _refuse("string", raw) from e


def _parse_int(raw: bytes) -> int:
    s = _ascii("int", raw)
    if not _INT.fullmatch(s):
        raise _refuse("int", raw)
    return int(s)


def _parse_decimal(raw: bytes) -> Decimal:
    s = _ascii("decimal", raw)
    if s == "-0" or not _DECIMAL.fullmatch(s):
        raise _refuse("decimal", raw)
    return Decimal(s)


def _parse_float(raw: bytes) -> float:
    s = _ascii("float", raw)
    if not _FLOAT.fullmatch(s):
        raise _refuse("float", raw)
    x = float(s)
    if not math.isfinite(x) or render_float(x) != raw:
        raise _refuse("float", raw)
    return x


def _parse_boolean(raw: bytes) -> bool:
    if raw == b"true":
        return True
    if raw == b"false":
        return False
    raise _refuse("boolean", raw)


def _parse_date(raw: bytes) -> dt.date:
    s = _ascii("date", raw)
    if not _DATE.fullmatch(s):
        raise _refuse("date", raw)
    try:
        return dt.date.fromisoformat(s)
    except ValueError as e:
        raise _refuse("date", raw) from e


def _parse_datetime(raw: bytes) -> dt.datetime:
    s = _ascii("datetime", raw)
    if not _DATETIME.fullmatch(s):
        raise _refuse("datetime", raw)
    try:
        return dt.datetime(int(s[0:4]), int(s[5:7]), int(s[8:10]),
                           int(s[11:13]), int(s[14:16]), int(s[17:19]),
                           int(s[20:26]), tzinfo=dt.UTC)
    except ValueError as e:  # year 0000, hour 24, second 60, February 30
        raise _refuse("datetime", raw) from e


_RENDER = {
    "string": _render_string, "bytes": _render_bytes, "int": _render_int,
    "decimal": _render_decimal, "float": _render_float,
    "boolean": _render_boolean, "date": _render_date,
    "datetime": _render_datetime,
}
_PARSE = {
    "string": _parse_string, "bytes": lambda raw: raw, "int": _parse_int,
    "decimal": _parse_decimal, "float": _parse_float,
    "boolean": _parse_boolean, "date": _parse_date,
    "datetime": _parse_datetime,
}


def _require_type(field: Field[Any, Any]) -> str:
    kind = logical_type(field)
    if kind is None:  # Encrypted() refuses these at declaration; defensive
        raise FieldsealNotSupported(unmapped_message(field))
    return kind


def unmapped_message(field: Field[Any, Any]) -> str:
    return (
        f"{type(field).__name__} has no spec §3.6 rendering. The pinned "
        "logical types are string (CharField, TextField and subclasses), "
        "bytes (BinaryField), int (IntegerField family), decimal "
        "(DecimalField), float (FloatField), boolean (BooleanField), date "
        "(DateField) and datetime (DateTimeField). Store anything else as a "
        "CharField holding a rendering your application owns.")

File: adapters/prisma/src/codec.ts
/**
 * Value <-> bytes, and envelope <-> column.
 *
 * Two conversions, deliberately separated because they fail differently.
 *
 * **The codec** turns an application value into the plaintext bytes that get
 * encrypted, under the column's declared `as:` type -- one of spec §3.6's eight
 * logical types -- and back. The rendering is §3.6's, byte for byte, and the
 * `codec/` vector family pins it (`MANIFEST.adapter_files`). Before §3.6 it was
 * this file's own choice, and it disagreed with the Django adapter on four of
 * eight types (G25, #123): a date written by Django came back from here as an
 * instant and was rewritten as one, which Django then could not read.
 *
 * Both directions refuse rather than coerce. A value §3.6 does not admit is
 * refused on write; bytes that are not the canonical rendering are refused on
 * read, because the envelope was authentic and a coerced value would hide
 * that the row was written by something else. Two refusals are specific to
 * JavaScript and are stated in §3.6: a `datetime` whose microseconds this
 * platform's `Date` cannot hold, and a `date` that is not exactly at the UTC
 * midnight this adapter uses to represent a calendar date.
 *
 * **The storage form** turns the envelope into what the column holds: raw
 * bytes for `Bytes`, base64 ASCII for a `String` column carrying
 * `storage: "base64"`. A consumer handed the wrong one fails at the length
 * gate with an error pointing at the envelope rather than at the column.
 *
 * Never pickle, never revive a stored value into a live object: a deserializer
 * that can construct arbitrary objects turns a decryption boundary into a
 * code-execution boundary.
 */

import { FieldsealNotSupported } from "./errors.ts";
import type { EncryptedFieldDecl } from "./fieldmap.ts";

const INT = /^(0|-?[1-9][0-9]*)$/;
const DECIMAL_CANONICAL = /^-?(0|[1-9][0-9]*)(\.[0-9]*[1-9])?$/;
/** Decimal notation a caller may write: sign, digits, point, exponent. */
const DECIMAL_INPUT = /^([+-]?)([0-9]*)(?:\.([0-9]*))?(?:[eE]([+-]?[0-9]+))?$/;
const FLOAT = /^-?[0-9]+(\.[0-9]+)?(e[+-][0-9]+)?$/;
const DATE = /^([0-9]{4})-([0-9]{2})-([0-9]{2})$/;
const DATETIME = /^([0-9]{4})-([0-9]{2})-([0-9]{2})T([0-9]{2}):([0-9]{2}):([0-9]{2})\.([0-9]{6})Z$/;
const STRICT_UTF8 = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true });

/** Application value -> plaintext bytes, under the declared `as:` type. */
export function toBytes(value: unknown, decl: EncryptedFieldDecl, label: string): Uint8Array {
  const enc = (s: string) => Buffer.from(s, "ascii");
  switch (decl.valueType) {
    case "string":
      if (typeof value === "string") {
        // Buffer.from would write U+FFFD for an unpaired surrogate without a
        // word; §3.6 refuses it.
        const at = unpairedSurrogateAt(value);
        if (at >= 0) {
          throw refuseWrite(label, "string", `it contains an unpaired surrogate at index ${at}, which has no UTF-8 encoding`);
        }
        return Buffer.from(value, "utf8");
      }
      break;
    case "bytes":
      if (value instanceof Uint8Array) return value;
      break;
    case "int":
      if (typeof value === "bigint") return enc(value.toString());
      if (typeof value === "number" && Number.isInteger(value)) {
        if (!Number.isSafeInteger(value)) {
          throw refuseWrite(label, "int", `${value} is beyond 2^53, where a number may already have been rounded; pass a bigint`);
        }
        return enc(BigInt(value).toString());
      }
      break;
    case "decimal": {
      const text = decimalText(value);
      if (text === null) break;
      const canon = canonicalDecimal(text);
      if (canon === null) throw refuseWrite(label, "decimal", `${JSON.stringify(text)} is not a finite decimal`);
      return enc(canon);
    }
    case "float":
      if (typeof value === "number") {
        if (!Number.isFinite(value)) throw refuseWrite(label, "float", `${value} is not finite`);
        return enc(renderFloat(value));
      }
      break;
    case "boolean":
      if (typeof value === "boolean") return enc(value ? "true" : "false");
      break;
    case "date":
      if (value instanceof Date && !Number.isNaN(value.getTime())) {
        const iso = value.toISOString();
        if (!iso.endsWith("T00:00:00.000Z")) {
          throw refuseWrite(label, "date", `${iso} is not exactly UTC midnight, which is how this adapter represents a calendar date; it is refused rather than truncated to a day`);
        }
        if (!inYearRange(value)) throw refuseWrite(label, "date", `${iso} is outside years 0001-9999`);
        return enc(iso.slice(0, 10));
      }
      break;
    case "datetime":
      if (value instanceof Date && !Number.isNaN(value.getTime())) {
        if (!inYearRange(value)) throw refuseWrite(label, "datetime", `${value.toISOString()} is outside years 0001-9999`);
        const iso = value.toISOString(); // YYYY-MM-DDTHH:MM:SS.sssZ in range
        return enc(`${iso.slice(0, 23)}000Z`);
      }
      break;
  }
  throw new FieldsealNotSupported(
    `${label}: declared \`as: "${decl.valueType}"\` but was given ` +
      `${describe(value)}. The adapter refuses rather than coercing: the byte ` +
      `rendering is spec §3.6's, what a reader in another language must decode, ` +
      `and a coerced value decrypts cleanly into the wrong thing. Either write the ` +
      `declared type, or change \`as:\` -- which is a new plaintext encoding for ` +
      `every row already written.${hint(decl.valueType)}`,
  );
}

/** Plaintext bytes -> application value, under the declared `as:` type. */
export function fromBytes(plaintext: Uint8Array, decl: EncryptedFieldDecl, label: string): unknown {
  const buf = Buffer.from(plaintext);
  switch (decl.valueType) {
    case "bytes":
      return buf;
    case "string":
      try {
        return STRICT_UTF8.decode(buf);
      } catch {
        break;
      }
    case "int": {
      const s = buf.toString("latin1");
      if (!INT.test(s)) break;
      const n = Number(s);
      return Number.isSafeInteger(n) ? n : BigInt(s);
    }
    case "decimal": {
      const s = buf.toString("latin1");
      if (s === "-0" || !DECIMAL_CANONICAL.test(s)) break;
      return s;
    }
    case "float": {
      const s = buf.toString("latin1");
      if (!FLOAT.test(s)) break;
      const n = Number(s);
      if (!Number.isFinite(n) || renderFloat(n) !== s) break;
      return n;
    }
    case "boolean": {
      const s = buf.toString("latin1");
      if (s === "true") return true;
      if (s === "false") return false;
      break;
    }
    case "date": {
      const m = DATE.exec(buf.toString("latin1"));
      if (m === null) break;
      const d = utc(+m[1]!, +m[2]!, +m[3]!, 0, 0, 0, 0);
      if (d === null) break;
      return d;
    }
    case "datetime": {
      const m = DATETIME.exec(buf.toString("latin1"));
      if (m === null) break;
      const micros = m[7]!;
      if (!micros.endsWith("000")) {
        // §3.6: a Date holds milliseconds. Truncating .123456 to .123 is what
        // Node would do silently; refuse instead.
        throw new FieldsealNotSupported(
          `${label}: the stored instant carries microseconds (.${micros}) that a ` +
            `JavaScript Date cannot hold. Spec §3.6 refuses rather than truncating; ` +
            `the row was written by a platform with microsecond instants.`,
        );
      }
      const d = utc(+m[1]!, +m[2]!, +m[3]!, +m[4]!, +m[5]!, +m[6]!, +micros.slice(0, 3));
      if (d === null) break;
      return d;
    }
  }
  // The bytes decrypted -- so the key, the context and the commitment were all
  // right -- and then were not the canonical rendering of the declared type.
  // Returning a coerced value would hide it.
  throw new FieldsealNotSupported(
    `${label}: the decrypted value is not spec §3.6's canonical \`${decl.valueType}\` ` +
      `rendering. The envelope was authentic, so this is not tampering: the row ` +
      `was written by something that does not render §3.6, or the column's ` +
      `declared type changed after it was written. Changing \`as:\` is a new ` +
      `plaintext encoding and needs a backfill, not an edit.`,
  );
}

/**
 * Spec §3.6 `float`: ECMA-262 Number::toString, which `String()` is, except
 * that negative zero keeps its sign -- `String(-0)` is `"0"`.
 */
function renderFloat(x: number): string {
  return Object.is(x, -0) ? "-0" : String(x);
}

/** The text of a decimal the caller handed over, or null for a wrong type. */
function decimalText(value: unknown): string | null {
  if (typeof value === "string") return value;
  // Prisma.Decimal (decimal.js): toFixed() with no argument is its exact,
  // exponent-free form. A number is refused: it is a binary64 already.
  if (
    typeof value === "object" && value !== null &&
    typeof (value as { toFixed?: unknown }).toFixed === "function" &&
    typeof (value as { isFinite?: unknown }).isFinite === "function"
  ) {
    const d = value as { toFixed(): string; isFinite(): boolean };
    return d.isFinite() ? d.toFixed() : "NaN";
  }
  return null;
}

/**
 * Spec §3.6 canonical decimal, by string arithmetic alone -- no binary64 ever
 * touches the digits. Null when `text` is not decimal notation or not finite.
 */
function canonicalDecimal(text: string): string | null {
  const m = DECIMAL_INPUT.exec(text);
  if (m === null) return null;
  const [, sign, ip = "", fp = "", e = "0"] = m;
  if (ip === "" && fp === "") return null; // ".", "", "e5"
  let digits = (ip + fp).replace(/^0+/, "");
  if (digits === "") return "0";
  let exp = Number(e) - fp.length; // value = digits * 10^exp
  if (!Number.isSafeInteger(exp)) return null;
  const stripped = digits.replace(/0+$/, "");
  exp += digits.length - stripped.length;
  digits = stripped;
  let body: string;
  if (exp >= 0) {
    body = digits + "0".repeat(exp);
  } else {
    const point = digits.length + exp;
    body = point > 0 ? `${digits.slice(0, point)}.${digits.slice(point)}` : `0.${"0".repeat(-point)}${digits}`;
  }
  return (sign === "-" ? "-" : "") + body;
}

/**
 * Index of the first unpaired UTF-16 surrogate, or -1. By hand rather than
 * `String.prototype.isWellFormed`, which is ES2024 and outside this package's
 * `lib`; the TypeScript core scans the same way (`normalize.ts`).
 */
function unpairedSurrogateAt(s: string): number {
  for (let i = 0; i < s.length; i++) {
    const unit = s.charCodeAt(i);
    if (unit >= 0xd800 && unit <= 0xdbff) {
      const next = s.charCodeAt(i + 1);
      if (next >= 0xdc00 && next <= 0xdfff) {
        i++;
        continue;
      }
      return i;
    }
    if (unit >= 0xdc00 && unit <= 0xdfff) return i;
  }
  return -1;
}

function inYearRange(d: Date): boolean {
  const y = d.getUTCFullYear();
  return y >= 1 && y <= 9999;
}

/** A UTC Date from components, or null unless every component survives. */
function utc(y: number, mo: number, d: number, h: number, mi: number, s: number, ms: number): Date | null {
  if (y < 1) return null;
  const out = new Date(0);
  // setUTCFullYear, not Date.UTC: Date.UTC maps years 0-99 to 1900-1999.
  out.setUTCFullYear(y, mo - 1, d);
  out.setUTCHours(h, mi, s, ms);
  // A rollover (February 30, hour 24, second 60) changes a component.
  const ok =
    out.getUTCFullYear() === y && out.getUTCMonth() === mo - 1 && out.getUTCDate() === d &&
    out.getUTCHours() === h && out.getUTCMinutes() === mi && out.getUTCSeconds() === s;
  return ok ? out : null;
}

function refuseWrite(label: string, type: string, why: string): FieldsealNotSupported {
  return new FieldsealNotSupported(`${label}: \`as: "${type}"\` refuses this value: ${why} (spec §3.6).`);
}

function hint(type: string): string {
  if (type === "decimal") return " A decimal is written as a string or a Prisma.Decimal, and read back as its canonical string.";
  if (type === "date") return " A date is written and read as a Date at exactly UTC midnight.";
  return "";
}

/** Envelope -> the value written to the column. */
export function toColumn(envelope: Uint8Array, decl: EncryptedFieldDecl): Uint8Array | string {
  return decl.storage === "base64"
    ? Buffer.from(envelope).toString("base64")
    : Buffer.from(envelope);
}

/** Column value -> envelope bytes, or `null` if this does not look stored. */
export function fromColumn(stored: unknown, decl: EncryptedFieldDecl): Uint8Array | null {
  if (stored === null || stored === undefined) return null;
  if (decl.storage === "base64") {
    return typeof stored === "string" ? Buffer.from(stored, "base64") : null;
  }
  return stored instanceof Uint8Array ? Buffer.from(stored) : null;
}

function describe(v: unknown): string {
  if (v === null) return "null";
  if (v === undefined) return "undefined";
  if (Array.isArray(v)) return "an array";
  if (v instanceof Uint8Array) return "bytes";
  return typeof v === "object" ? `a ${v.constructor?.name ?? "object"}` : `a ${typeof v}`;
}
